css/css-view-transitions/pseudo-element-animations fails to remove animations when the document element gets display:none.

mattwoodrow opened the pull request:
#### ca5554a142fd1f3bbe12b7161113e5d8359dc534
<pre>
css/css-view-transitions/pseudo-element-animations fails to remove animations when the document element gets display:none.
<a href="https://bugs.webkit.org/show_bug.cgi?id=293034">https://bugs.webkit.org/show_bug.cgi?id=293034</a>
&lt;<a href="https://rdar.apple.com/151353389">rdar://151353389</a>&gt;

Reviewed by Tim Nguyen.

Setting display:none on the document element should remove the VT tree (and all associated animations), but does not because the VT renderer tree is a sibling of the document element&apos;s renderer.

Make sure RenderTreeUpdater explicitly clears the VT tree if the document element renderer gets removed, and make sure we remove animations (on removed renderers, and their descendants).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::clearViewTransition):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateRenderTree):
(WebCore::RenderTreeUpdater::updateAfterDescendants):
(WebCore::RenderTreeUpdater::destroyAndCancelAnimationsForSubtree):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h:

Canonical link: <a href="https://commits.webkit.org/295170@main">https://commits.webkit.org/295170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/152ea9a5f0ef3cfff528ee4ea21c4dc6c351e9db

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104213 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23917 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109413 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54881 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106253 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32462 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79150 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59477 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18682 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/postmessage-to-client-message-queue.https.html (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12058 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54241 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88382 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12117 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111795 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31369 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23124 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88178 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31733 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90209 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87839 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22376 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32726 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10475 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25833 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31298 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36611 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31092 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34428 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32652 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -358,8 +360,11 @@ void RenderTreeUpdater::updateAfterDescendants(Element& element, const Style::El
     auto* renderer = element.renderer();
-    if (!renderer)
+    if (!renderer) {
+        if (&element == element.document().documentElement())
+            viewTransition().updatePseudoElementTree(nullptr, StyleDifference::Equal);
```
not sure how hot this is, but maybe worth A/B testing perf?

nt1m reviewed (approved):
The approach seems OK, and probably the least invasive one.

mattwoodrow committed:
css/css-view-transitions/pseudo-element-animations fails to remove animations when the document element gets display:none.
https://bugs.webkit.org/show_bug.cgi?id=293034
<rdar://151353389>

Reviewed by Tim Nguyen.

Setting display:none on the document element should remove the VT tree (and all associated animations), but does not because the VT renderer tree is a sibling of the document element's renderer.

Make sure RenderTreeUpdater explicitly clears the VT tree if the document element renderer gets removed, and make sure we remove animations (on removed renderers, and their descendants).

* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/pseudo-element-animations-expected.txt:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::clearViewTransition):
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateRenderTree):
(WebCore::RenderTreeUpdater::updateAfterDescendants):
(WebCore::RenderTreeUpdater::destroyAndCancelAnimationsForSubtree):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementTree):
(WebCore::RenderTreeUpdater::ViewTransition::updatePseudoElementGroup):
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.h:

Canonical link: https://commits.webkit.org/295170@main
